### PR TITLE
[ROX-9744][ROX-9437]: Marking 'whitelists' field in policy proto as reserved.  Deprecating PolicyFields. Removing a bunch of code paths related to policy versions, upgrade and downgrade.

### PR DESCRIPTION
**Pull request by clickboo**
## Description
1.  Addressing code paths that are now obsolete and removing them.
2. Deprecating the use of whitelists and PolicyFields by marking those reserved.
3. Ensuring only policy versions 1.1 are supported, and removing any code paths that allow upgrade, downgrade to other versions (note that the mechanism to upgrade/downgrade is left as-is). After this PR, only one version will be recognized by our product - policy version 1.1.

## Checklist
- [x] Investigated and inspected CI test results
- [x] Unit test and regression tests added
- [x] Evaluated and added CHANGELOG entry if required
- [x] ~Determined and documented upgrade steps~
- [x] Documented user facing changes (create PR based on [stackrox/openshift-docs](https://github.com/stackrox/openshift-docs) and merge into [rhacs-docs](https://github.com/openshift/openshift-docs/tree/rhacs-docs)).

If any of these don't apply, please comment below.

## Testing Performed

TODO(replace-me)
Use this space to explain how you tested your PR, or, if you didn't test it, why you did not do so. (Valid reasons include "CI is sufficient" or "No testable changes")
In addition to reviewing your code, reviewers **must** also review your testing instructions, and make sure they are sufficient.

For more details, ref the [Confluence page](https://stack-rox.atlassian.net/wiki/spaces/StackRox/pages/855998488/Proposal+Explicitly+List+Testing+Steps+on+PRs) about this section.

### CODE REVIEW

**Review by connorgorman — DISMISSED** [review]
Seems reasonable as long as product has signed off on removing the backwards compatibility for policy versions

**Review by md2119 — DISMISSED** [review]
Merge with changelog

File: pkg/booleanpolicy/policyversion/downgrade_test.go [hunk 1]
```diff
@@ -143,26 +66,24 @@ func TestOrganizeByVersionRank(t *testing.T) {
...
+				version1_1: dummy,
 			},
 			[]downgrader{
 				dummy, dummy, dummy,
 			},
 		},
 	}
 
-	versions := []string{legacyVersion, version1, version1_1}
+	versions := []string{version1_1}
```
[the remark this answers is not in the document]
Comment (clickboo) [reply pkg/booleanpolicy/policyversion/downgrade_test.go:86]: I do not have a crystal ball for the future :-) though I wish I did. My goal is to remove unnecessary, unused code paths and reduce code bloat.  The workflows team has to navigate this day in an day out, as we develop our features, unnecessarily.

[the remark this answers is not in the document]
Comment (clickboo) [reply pkg/booleanpolicy/policyversion/downgrade_test.go:86]: > In its current state, this PR removes _parts_ of the versioning functionality (with _some_ tests) rendering the whole feature useless but still present in the code. This is suboptimal, I suggest to take one of these 3 options:
> 
> 1. remove versioning in backend entirely, including all upgrade, downgrade, version compare code and tests;
> 2. same as ^ but also remove `Policy.policy_version` proto field;
> 3. remove the possibility to downgrade from current `1.1` version to anything below but keep all the versioning functionality for future need.
> 
> My intuition says that since we've already seen 2 cases where we needed to version our policies, it's hard to believe that no more cases will arise, so 3. seems to be the best option. However, I'm not familiar with the vision for our policy engine and hence might be off.
> 
> Let's agree (and capture it in the PR description) on what we want here before we resume reviewing.

Agree. 3 has my vote. Remove the complexity, and unnecessary baggage once and for all :-). My latest commit is towards that, check it out when you have a few!

File: pkg/booleanpolicy/policyversion/convertor.go [hunk 2]
```diff
@@ -11,52 +11,10 @@ func EnsureConvertedToLatest(p *storage.Policy) error {
...
-			return errors.New("both 'exclusions' and 'whitelists' fields are set")
-		}
-
-		upgradeVersion1ToVersion1_1(p)
-	}
-
-	if p.PolicyVersion != CurrentVersion().String() {
-		return errors.Errorf("converted to version %q, while latest is %q", p.PolicyVersion, CurrentVersion().String())
-	}
+	p.PolicyVersion = CurrentVersion().String()
```
[the remark this answers is not in the document]
Comment (clickboo) [reply pkg/booleanpolicy/policyversion/convertor.go:14]: Yes, all policies from here on will have policy version 1.1 stamped in them. Anything else won't work.

File: pkg/booleanpolicy/policyversion/downgrade_test.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (rukletsov) [reply pkg/booleanpolicy/policyversion/downgrade_test.go]: I'm confused. You say 3 but describe 1. What shall I review?

File: pkg/booleanpolicy/policyversion/downgrade_test.go [hunk 3]
```diff
@@ -4,27 +4,10 @@ import (
...
 	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/require"
 )
 
-// If this test fails, chances are you have added a new policy version have not
-// provided a downgrader from this new version to the previous one. Consider
-// doing so. If it is impossible to downgrade from the new version, add an
-// exception here.
-func TestNoDowngraderLeftBehind(t *testing.T) {
-	noDowngradersFromVersions := set.NewStringSet(legacyVersion, version1)
```
[the remark this answers is not in the document]
Comment (clickboo) [reply pkg/booleanpolicy/policyversion/downgrade_test.go:17]: Removed the file.

File: pkg/booleanpolicy/policyversion/downgrade_test.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (clickboo) [reply pkg/booleanpolicy/policyversion/downgrade_test.go]: Removed the downgrade path (please check latest commit when you get a chance).

File: central/processbaseline/service/service.go [hunk 4]
```diff
@@ -22,9 +22,6 @@ type Service interface {
 	AuthFuncOverride(ctx context.Context, fullMethodName string) (context.Context, error)
 
 	v1.ProcessBaselineServiceServer
-
-	// TODO(ROX-6194): Remove after the deprecation cycle started with the 55.0 release.
-	v1.ProcessWhitelistServiceServer
```
[the remark this answers is not in the document]
Comment (clickboo) [reply central/processbaseline/service/service.go:27]: As discussed why needed (this was a request from Gorman, check comments above)

**Comment by clickboo** [comment]
> All three failing CI jobs say that sensor can't start — doubt this is a coincidence.

https://srox.slack.com/archives/CELUQKESC/p1649208113384909

File: pkg/booleanpolicy/policyversion/version.go [hunk 5]
```diff
@@ -37,12 +36,6 @@ func CurrentVersion() PolicyVersion {
 	return PolicyVersion{versions[len(versions)-1]}
```
Comment (rukletsov) [inline pkg/booleanpolicy/policyversion/version.go:36]: Regardless on whether we want to keep the policy check in the Central<->Sensor communication, most of this file, except maybe `CurrentVersion()`, can go away.

File: pkg/booleanpolicy/policyversion/version_test.go [hunk 6]
```diff
@@ -46,48 +45,10 @@ func TestVersionCompare(t *testing.T) {
 	}
 
 	testCases := []TestCase{
```
Comment (rukletsov) [inline pkg/booleanpolicy/policyversion/version_test.go:47]: No need for any tests if versioning goes away.

File: pkg/booleanpolicy/policyversion/convertor.go [hunk 7]
```diff
@@ -1,67 +1,19 @@
...
+	"fmt"
+
 	"github.com/pkg/errors"
 	"github.com/stackrox/rox/generated/storage"
-	"github.com/stackrox/rox/pkg/utils"
 )
 
-// EnsureConvertedToLatest converts the given policy into a Boolean policy, if it is not one already.
+// EnsureConvertedToLatest converts the given policy into a Boolean policy version 1.1, if it is not one already.
 func EnsureConvertedToLatest(p *storage.Policy) error {
```
Comment (rukletsov) [inline pkg/booleanpolicy/policyversion/convertor.go:11]: Why would we keep this if we remove all versioning stuff?

File: pkg/booleanpolicy/validate.go [hunk 8]
```diff
@@ -48,9 +47,6 @@ func Validate(p *storage.Policy, options ...ValidateOption) error {
 	}
 
 	errorList := errorhelpers.NewErrorList("policy validation")
-	if !policyversion.IsBooleanPolicy(p) {
```
Comment (rukletsov) [inline pkg/booleanpolicy/validate.go:51]: I'd check `policy_version` is "1.1", assuming you plan to keep this proto field.

**Review by rukletsov — CHANGES_REQUESTED** [review]
In addition please consider removing the following:
* `tools/policyutil` 
* `check-policy-files-up-to-date` CI stage

File: CHANGELOG.md [hunk 9]
```diff
@@ -60,8 +61,7 @@ Please avoid adding duplicate information across this changelog and JIRA/doc inp
...
 - Vulnerability snoozing and un-snoozing will not impact image and component risk. Furthermore, it will not impact `Image Vulnerabilities` risk factor for deployments.
-- Note that this only affects users who are storing their policies externally and keeping those policies in sync with Central. There is no action required for users who are not taking this approach.
-  - Support for system policies with `policyVersion` unset will be removed in 3.71. All such externally stored policies must be converted to policy version 1.0 or higher. To do so, import the policies into the system, and then export the previously imported policies. The exported policies will now be in supported policy version format (>=1.0), which can be ensured by checking the `policyVersion` field.
+- In 3.70, support for security policies that do not have a policyVersion will be removed. Therefore, if you have externally stored older policies (without policyVersion or version prior to 1.1), you must convert them to use policyVersion 1.1. To do this, import the old policies into RHACS and then export them again. You can check the policyVersion field for your stored policies to identify if they need conversion.
```
Comment (theencee) [inline CHANGELOG.md:64]: I copied this from the actual release notes that shipped with 69.1 as it is slightly different from what the CHANGELOG originally had.

File: central/sensor/service/connection/connection_impl.go [hunk 10]
```diff
@@ -290,24 +290,29 @@ func (c *sensorConnection) getPolicySyncMsg(ctx context.Context) (*central.MsgTo
...
 	// Forward policies as is if we don't understand sensor's version.
 	if sensorPolicyVersion, err := policyversion.FromString(sensorPolicyVersionStr); err != nil {
-		log.Errorf("Cannot understand sensor's policy version %q: %v", sensorPolicyVersionStr, err)
+		log.Errorf("Cannot understand sensor's policy version %q, will assume version %q: %v", sensorPolicyVersionStr,
+			policyversion.CurrentVersion().String(), err)
 	} else {
-		// Downgrade all policies if necessary. If we can't downgrade one,
-		// we likely can't convert any of them, so no need to spam the log.
-		if policyversion.Compare(policyversion.CurrentVersion(), sensorPolicyVersion) > 0 {
+		if policyversion.Compare(sensorPolicyVersion, policyversion.MinimumSupportedVersionForSensor()) < 0 {
```
Comment (theencee) [inline central/sensor/service/connection/connection_impl.go:308]: this is new

File: pkg/booleanpolicy/policyversion/downgrade_test.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (theencee) [reply pkg/booleanpolicy/policyversion/downgrade_test.go:17]: I made this whole file generic and agnostic to what version we're actually in

File: pkg/booleanpolicy/policyversion/upgrade.go [hunk 11]
```diff
@@ -1,643 +1,67 @@
...
-	return []*storage.PolicyGroup{{
-		FieldName: fieldnames.WritableHostMount,
-		Values:    getPolicyValues(!p.GetReadOnly()),
-	},
-	}
-}
-
-const (
-	capPrefix = "CAP_"
+// Upgrader takes a policy in version N and *must* convert it to N+1.
```
Comment (theencee) [inline pkg/booleanpolicy/policyversion/upgrade.go:8]: This file has not been deleted! Hopefully this comment will force GitHub to render it.

It's been massively trimmed down by deleting anything that hd to do with fields or old versions. It simply has the same interface as downgrade.go now. But it doesn't do anything since there's only one version (1.1) that's active.

**Review by theencee — COMMENTED** [review]
Re-requesting review since I made a few changes to address comments. Two of the biggest things that changed are:

* Rebased (there were a lot of changes)
* Brought back the upgrade and downgrade code but only as a stub. It should be straightforward to add a new upgrader or downgrader in the future, but it doesn't have any now. I kept the interface on advice that `pkg` has code that is intended to be a library that may be used in the future. The old upgrader code was very specific to policy fields, so I rewrote it similar to downgrader so that it's generic.

Please take a look and let me know if I missed something

File: central/policy/datastore/datastore_impl.go [hunk 12]
```diff
@@ -137,6 +138,9 @@ func (ds *datastoreImpl) AddPolicy(ctx context.Context, policy *storage.Policy)
 	// Any policy added after statup must be marked custom policy.
 	markPoliciesAsCustom(policy)
 
+	// Set version
+	policy.PolicyVersion = policyversion.CurrentVersion().String()
+
```
Comment (viswajithiii) [inline central/policy/datastore/datastore_impl.go:143]: This should not be done here. 🤔 What made this necessary suddenly?

File: pkg/booleanpolicy/policyversion/convertor.go [hunk 13]
```diff
@@ -6,40 +6,24 @@ import (
...
 	}
-	policyVersion, err := FromString(p.GetPolicyVersion())
-	if err != nil {
-		return err
-	}
 
-	if Compare(policyVersion, Version1()) >= 0 && len(p.GetPolicySections()) == 0 {
-		return errors.New("empty sections")
+	ver, err := FromString(p.GetPolicyVersion())
+	if err != nil || ver.String() == legacyVersion { // Legacy versions (i.e. no version) is no longer supported
```
Comment (viswajithiii) [inline pkg/booleanpolicy/policyversion/convertor.go:17]: I don't think this `ver.String() == legacyVersion` comparison is required here. This can just be allowed to happen through the error returned by `UpgradePolicyTo` below.

File: pkg/booleanpolicy/policyversion/upgrade.go [hunk 14]
```diff
@@ -1,643 +1,67 @@
...
-	perPolicy := fields.GetPermissionPolicy()
-	if perPolicy == nil {
-		return nil
-	}
-
-	permissionLevel, err := getRBACPermissionName(perPolicy.GetPermissionLevel())
+// UpgradePolicyTo attempts to upgrade a given policy to the policy in the
+// given target version. Policies in some versions cannot be upgraded (for e.g. one that is already at the latest).
+// The function leaves policy either unchanged or in the upgraded state.
+func UpgradePolicyTo(p *storage.Policy, targetVersion PolicyVersion) error {
```
Comment (viswajithiii) [inline pkg/booleanpolicy/policyversion/upgrade.go:28]: This function doesn't need to be exported?

**Review by viswajithiii — APPROVED** [review]
This seems fine except for the comments below.

I also feel there is some documentation debt here, which is probably exacerbated by this PR (since we've removed existing upgraders/downgraders, so now people will have no existing ones to emulate). We should have a documented guide for how to update policy versions so people know what the steps they need to take are (including writing downgraders if possible).

File: central/compliance/checks/testutils/policy.go [hunk 15]
```diff
@@ -30,24 +33,61 @@ func (l *LightPolicy) convert() *storage.Policy {
...
+				{
+					Value: fmt.Sprintf("=%s=%s", l.EnvKey, l.EnvValue),
+				},
+			},
+		})
 	}
 	if l.Enforced {
 		p.EnforcementActions = append(p.EnforcementActions, storage.EnforcementAction_SCALE_TO_ZERO_ENFORCEMENT)
 	}
+	p.PolicyVersion = policyversion.CurrentVersion().String()
```
Comment (md2119) [inline central/compliance/checks/testutils/policy.go:86]: why was this necessary?

File: central/compliance/checks/testutils/policy.go [hunk 16]
```diff
@@ -30,24 +33,61 @@ func (l *LightPolicy) convert() *storage.Policy {
...
 	if err := policyversion.EnsureConvertedToLatest(p); err != nil {
```
Comment (md2119) [inline central/compliance/checks/testutils/policy.go:87]: The above is marking the policy with latest version. Isn't calling`EnsureConvertedToLatest` moot?

File: central/policy/service/validator.go [hunk 17]
```diff
@@ -104,17 +105,24 @@ func (s *policyValidator) internalValidate(policy *storage.Policy, additionalVal
...
-	if !policyversion.IsBooleanPolicy(policy) {
-		return errors.New("policy not converted to boolean policy")
+	ver, err := policyversion.FromString(policy.GetPolicyVersion())
+	if err != nil {
+		return errors.New("policy has invalid version")
+	}
+
+	// As of 70.0 we only support the latest version (1.1). This may in the future be expanded to support more, but for
+	// now it's enough to just check it matches current version
+	if !policyversion.IsCurrentVersion(ver) {
```
Comment (md2119) [inline central/policy/service/validator.go:116]: Can we change this to `IsSupportedVersion()`? The supported versions need not be current version.

File: central/processbaseline/service/service_impl.go [hunk 18]
```diff
@@ -27,19 +27,11 @@ var (
...
 		},
 		user.With(permissions.Modify(resources.ProcessWhitelist)): {
 			"/v1.ProcessBaselineService/UpdateProcessBaselines",
 			"/v1.ProcessBaselineService/LockProcessBaselines",
 			"/v1.ProcessBaselineService/DeleteProcessBaselines",
-
-			// TODO(ROX-6194): Remove after the deprecation cycle started with the 55.0 release.
-			"/v1.ProcessWhitelistService/UpdateProcessWhitelists",
-			"/v1.ProcessWhitelistService/LockProcessWhitelists",
-			"/v1.ProcessWhitelistService/DeleteProcessWhitelists",
```
Comment (md2119) [inline central/processbaseline/service/service_impl.go:42]: Why is this diff here?

File: proto/api/v1/process_whitelist_service.proto [hunk 19]
```diff
@@ -1,96 +0,0 @@
-syntax = "proto3";
```
Comment (md2119) [inline proto/api/v1/process_whitelist_service.proto:1]: Was a changelog ever published for this removal?

File: proto/storage/policy.proto [hunk 20]
```diff
@@ -20,12 +20,12 @@ message Policy {
     string                      remediation             = 5;
     bool                        disabled                = 6 [(gogoproto.moretags) = "search:\"Disabled\""];
     repeated string             categories              = 7 [(gogoproto.moretags) = "search:\"Category,store\""];
-    PolicyFields                fields                  = 8 [(gogoproto.moretags) = 'sql:"-"'];
+    reserved                                              8; // was PolicyFields fields
```
Comment (md2119) [inline proto/storage/policy.proto:23]: I don't see a reason why we should keep `PolicyFields` protobuf. Let's remove the it altogether.

Comment (md2119) [reply proto/storage/policy.proto:23]: I am not sure why you pointed to https://github.com/stackrox/stackrox/pull/879#issuecomment-1063525408.

Thus far, this PR has only removed the usage of the `PolicyFields` in the `Policy` protobuf. Additionally, we should remove the definition of `PolicyFields` aka `PolicyFields` protobuf.

File: central/processbaseline/service/service_impl.go (hunk 18 above)
Comment (md2119) [reply central/processbaseline/service/service_impl.go:42]: The diff I pointed out above is for process baseline APIs. Why are we removing it in policy PR?

Comment (theencee) [reply central/processbaseline/service/service_impl.go:42]: I'm not sure I understand. The functions removed are for Process **Whitelist** service which has been deprecated since 55. Process **Baseline** are still around

File: proto/api/v1/process_whitelist_service.proto (hunk 19 above)
Comment (theencee) [reply proto/api/v1/process_whitelist_service.proto:1]: I'm not sure I understand. The file removed are for Process **Whitelist** service which has been deprecated since 55. Process **Baseline** are still around

File: central/processbaseline/service/service_impl.go (hunk 18 above)
Comment (theencee) [reply central/processbaseline/service/service_impl.go:42]: If you're question is why was this combined with a PR that touches policy fields.... good question. Unfortunately that ship sailed. Going forward something like this should not have been combined in the first place

**Review by rukletsov — APPROVED** [review]
I left only a few minor comments and nits, the change looks good overall. Thanks, Nakul.
